apis/alarm/src/lib.rs

```diff
@@ -0,0 +1,109 @@
+#![no_std]
+
+use libtock_platform as platform;
+use libtock_platform::share;
+use libtock_platform::{DefaultConfig, ErrorCode, Syscalls};
+
+/// The alarm driver
+///
+/// # Example
+/// ```ignore
+/// use libtock2::Alarm;
+///
+/// // Wait for timeout
+/// Alarm::sleep(Alarm::Milliseconds(2500));
+/// ```
+
+pub struct Alarm<S: Syscalls, C: platform::subscribe::Config = DefaultConfig>(S, C);
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct Hz(pub u32);
+
+pub trait Convert {
+    fn to_ticks(self, freq: Hz) -> Ticks;
+}
+
+#[derive(Copy, Clone, Debug)]
+pub struct Ticks(pub u32);
+
+impl Convert for Ticks {
+    fn to_ticks(self, _freq: Hz) -> Ticks {
+        self
+    }
+}
+
+#[derive(Copy, Clone)]
+pub struct Milliseconds(pub u32);
+
+impl Convert for Milliseconds {
+    fn to_ticks(self, freq: Hz) -> Ticks {
+        // Saturating multiplication will top out at about 1 hour at 1MHz.
+        // It's large enough for an alarm, and much simpler than failing
+        // or losing precision for short sleeps.
+        Ticks(self.0.saturating_mul(freq.0) / 1000)
+    }
+}
+
+impl<S: Syscalls, C: platform::subscribe::Config> Alarm<S, C> {
+    /// Run a check against the console capsule to ensure it is present.
+    ///
+    /// Returns number of concurrent notifications supported,
+    /// 0 if unbounded.
+    #[inline(always)]
+    pub fn driver_check() -> Result<u32, ErrorCode> {
+        S::command(DRIVER_NUM, command::DRIVER_CHECK, 0, 0).to_result()
+    }
+
+    pub fn get_frequency() -> Result<Hz, ErrorCode> {
+        S::command(DRIVER_NUM, command::FREQUENCY, 0, 0)
+            .to_result()
+            .map(Hz)
+    }
+
+    pub fn sleep_for<T: Convert>(time: T) -> Result<(), ErrorCode> {
+        let freq = Self::get_frequency()?;
+        let ticks = time.to_ticks(freq);
+
+        let called = core::cell::Cell::new(Option::<(u32, u32)>::None);
+        share::scope(|subscribe| {
+            S::subscribe::<_, _, C, DRIVER_NUM, { subscribe::CALLBACK }>(subscribe, &called)?;
+
+            S::command(DRIVER_NUM, command::SET_RELATIVE, ticks.0, 0)
+                .to_result()
+                .map(|_when: u32| ())?;
```

Comment:
I'm thinking about the edge cases here: IIRC our alarm API does not specify that it may invoke the callback earlier than requested. Still that might be the case implicitly, for when the timer resolution is insufficient. Due to scheduling artifacts and overhead, the callback may arrive delayed. Perhaps we should document that this function blocks for the desired time or more, modulo the time resolution?

Reply:
On most platforms, `sleep` will never return early, but may return late. I don't think that Tock wants to be unique in that respect.

Maybe we should change the alarm API and implementations to make sure they sleep at least as long as specified?

Reply:
Specifically, I think all we would need to do to ensure this in libtock-rs is specify that `to_ticks()` will always round up